tests/test_callback_for.py

```diff
@@ -0,0 +1,64 @@
+import scrapy
+import pytest
+from scrapy.utils.reqser import request_to_dict, request_from_dict
+
+from scrapy_po.webpage import ItemPage, WebPage
+from scrapy_po.utils import callback_for, DummyResponse
+
+
+class FakePage(ItemPage):
+
+    def to_item(self):
+        return 'it works!'
+
+
+class MySpider(scrapy.Spider):
+
+    name = 'my_spider'
+    parse_item = callback_for(ItemPage)
+    parse_web = callback_for(WebPage)
+
+
+def test_callback_for():
+    """Simple test case to ensure it works as expected."""
+    cb = callback_for(FakePage)
+    assert callable(cb)
+
+    fake_page = FakePage()
+    response = DummyResponse('http://example.com/')
+    assert list(cb(response=response, page=fake_page)) == ['it works!']
+
+
+def test_callback_serialization():
+    """Make sure callbacks are serializable."""
+    spider = MySpider()
+
+    # Sample request without callback (fallback to default spider.parse)
+    request = scrapy.Request('http://example.com/')
+    request_dict = request_to_dict(request, spider)
+    assert isinstance(request_dict, dict)
+    assert request_dict['url'] == 'http://example.com/'
+    assert request_dict['callback'] is None
+
+    # Sample request referencing callback on spider
+    request = scrapy.Request('http://example.com/', callback=spider.parse_item)
+    request_dict = request_to_dict(request, spider)
+    assert isinstance(request_dict, dict)
+    assert request_dict['url'] == 'http://example.com/'
+    assert request_dict['callback'] == 'parse_item'
```

Comment:
Hey @kmike, this assertion is failing because Scrapy is failing to find the proper parser name. `request_dict['callback']` is returning `'parse'` here because that's the name of the method we're returning from the `callback_for` function.

Currently, Scrapy is executing this code to find the callback name:

```python
def _find_method(obj, func):
    if obj:
        try:
            func_self = func.__self__
        except AttributeError:  # func has no __self__
            pass
        else:
            if func_self is obj:
                name = func.__func__.__name__
                if _is_private_method(name):
                    return _mangle_private_name(obj, func, name)
                return name
    raise ValueError("Function %s is not a method of: %s" % (func, obj))
```

I'd like to refactor this method on Scrapy's repository. That will make it possible for us to use `callback_for` as a spider attribute. 

It's worth noting that the generated callback won't work as an inline callback. It has to be attributed to the spider.

Let me know if you have other thoughts.

Reply:
Here's the pull request https://github.com/scrapy/scrapy/pull/4500

Reply:
The Scrapy change makes sense, thanks @victor-torres.

Inline usage would be nice though. I wonder if we should change the approach - create a mixing for a spider, which provides callback_for method, and overrides `__getattr__`(?). 

Inline usage would be then `scrapy.Request(url, self.callback_for(ProductPage))`.

callback_for can generate some method name (e.g. `__page_object_callback_<ProductPage import path>`. `__getattr__` would ensure that when attribute with a name starting with `__page_object_callback_` is requested, a callback for this page object is created; page object class is probably loaded by scrapy.utils.misc.load_object. Generated callbacks are cached.

This is a hack, but it seems it can work in the meantime. Once we stabilize the whole thing, and organize it better, we can add direct pageobject callbacks support to Scrapy.

Reply:
If the hack is too hackish, we can skip it for now, and just document that inline callback_for doesn't work properly with disk queues, consuming the memory.

Reply:
Thank you for your suggestion. It's definitely a way of archiving this behavior. 

One thing that we might consider here is that we would be overwriting a Scrapy's behavior as it currently does not support inline callbacks with Disk Queues independent from scrapy-po.

Implementing this hack would bring inconsistency to our ecosystem. I believe we should discuss something like this on the Scrapy repo itself. Does it make sense?

Reply:
Fair enough. I don't have a generic proposal on how to fix it for Scrapy though. Any ideas? If any, please open a Scrapy ticket for that :)